File: custom_components/simple_irrigation/frontend/src/cycle.ts
/**
 * Cycle → schedule-slot generation, mirroring `cycle.py` exactly.
 *
 * Used for the live wizard preview; the identical rules run server-side in
 * `cycle.py` for `cycle_upsert`, so the preview and the created slots always
 * agree (acceptance §8.4). Cycle length is 7 or 14 days because `week_parity`
 * can only express a 2-week cycle.
 */

import { isoWeekNumber, type WeekParity } from "./timetable-model";

export type CycleKind =
  | "daily"
  | "twice_daily"
  | "every_n_days"
  | "n_per_week"
  | "weekly"
  | "biweekly"
  | "custom";

export interface CycleMeta {
  label?: string;
  n?: number;
  anchor_weekday?: number;
  times?: string[];
  week_days?: number[];
}

export interface CycleSlotSpec {
  weekdays: number[];
  time_local: string;
  week_parity: WeekParity;
}

/** Round half up (matches JS Math.round and Python `round_half_up`). */
export function roundHalfUp(x: number): number {
  return Math.floor(x + 0.5);
}

export function oppositeParity(p: WeekParity): WeekParity {
  if (p === "odd") return "even";
  if (p === "even") return "odd";
  return "every";
}

/** Monday-based weekday of a JS Date (0 = Monday … 6 = Sunday). */
export function mondayBasedWeekday(d: Date): number {
  return (d.getDay() + 6) % 7;
}

/** Parity of the ISO week containing the next occurrence of the anchor weekday. */
export function anchorWeekParity(anchorWeekday: number, today: Date): Exclude<WeekParity, "every"> {
  const a = Math.max(0, Math.min(6, Math.round(anchorWeekday)));
  for (let i = 0; i < 8; i++) {
    const d = new Date(today.getFullYear(), today.getMonth(), today.getDate() + i);
    if (mondayBasedWeekday(d) === a) {
      return isoWeekNumber(d) % 2 === 1 ? "odd" : "even";
    }
  }
  return "odd";
}

function times(meta: CycleMeta | undefined): string[] {
  const raw = meta?.times;
  const out: string[] = [];
  if (Array.isArray(raw)) {
    for (const x of raw) {
      const s = String(x).trim();
      if (s) out.push(s);
    }
  }
  return out.length ? out : ["06:00"];
}

function anchor(meta: CycleMeta | undefined): number {
  const n = Number(meta?.anchor_weekday ?? 0);
  return Number.isFinite(n) ? Math.max(0, Math.min(6, Math.round(n))) : 0;
}

function nValue(meta: CycleMeta | undefined, def = 2): number {
  const n = Number(meta?.n ?? def);
  return Number.isFinite(n) ? Math.max(1, Math.round(n)) : def;
}

function weekDays(meta: CycleMeta | undefined): number[] {
  const raw = meta?.week_days;
  const out: number[] = [];
  const seen = new Set<number>();
  if (Array.isArray(raw)) {
    for (const x of raw) {
      const v = Number(x);
      if (Number.isInteger(v) && v >= 0 && v <= 6 && !seen.has(v)) {
        seen.add(v);
        out.push(v);
      }
    }
  }
  out.sort((a, b) => a - b);
  return out;
}

function everyNDaysSlots(
  n: number,
  a: number,
  timeLocal: string,
  p0: Exclude<WeekParity, "every">
): CycleSlotSpec[] {
  const runs = Math.max(1, roundHalfUp(14 / n));
  const offsets: number[] = [];
  for (let k = 0; k < runs; k++) offsets.push(roundHalfUp((k * 14) / runs));

  const weekA: number[] = [];
  const weekB: number[] = [];
  const seenA = new Set<number>();
  const seenB = new Set<number>();
  for (const off of offsets) {
    const wd = (a + off) % 7;
    if (off < 7) {
      if (!seenA.has(wd)) {
        seenA.add(wd);
        weekA.push(wd);
      }
    } else if (!seenB.has(wd)) {
      seenB.add(wd);
      weekB.push(wd);
    }
  }
  weekA.sort((x, y) => x - y);
  weekB.sort((x, y) => x - y);

  const sameSet =
    weekA.length === weekB.length && weekA.every((v, i) => v === weekB[i]);
  if (weekA.length && weekB.length && sameSet) {
    return [{ weekdays: weekA, time_local: timeLocal, week_parity: "every" }];
  }
  const out: CycleSlotSpec[] = [];
  if (weekA.length) out.push({ weekdays: weekA, time_local: timeLocal, week_parity: p0 });
  if (weekB.length)
    out.push({ weekdays: weekB, time_local: timeLocal, week_parity: oppositeParity(p0) });
  return out;
}

export function generateCycleSlots(
  kind: CycleKind,
  meta: CycleMeta | undefined,
  anchorParity: Exclude<WeekParity, "every"> = "odd"
): CycleSlotSpec[] {
  const ts = times(meta);
  const a = anchor(meta);
  const allDays = [0, 1, 2, 3, 4, 5, 6];

  switch (kind) {
    case "daily":
      return [{ weekdays: allDays, time_local: ts[0], week_parity: "every" }];
    case "twice_daily": {
      const t2 = ts.length > 1 ? ts[1] : ts[0];
      return [
        { weekdays: allDays, time_local: ts[0], week_parity: "every" },
        { weekdays: allDays, time_local: t2, week_parity: "every" },
      ];
    }
    case "weekly":
      return [{ weekdays: [a], time_local: ts[0], week_parity: "every" }];
    case "biweekly":
      return [{ weekdays: [a], time_local: ts[0], week_parity: anchorParity }];
    case "n_per_week": {
      const days = weekDays(meta);
      return [
        { weekdays: days.length ? days : [a], time_local: ts[0], week_parity: "every" },
      ];
    }
    case "every_n_days":
      return everyNDaysSlots(nValue(meta), a, ts[0], anchorParity);
    default: {
      const days = weekDays(meta);
      return [
        { weekdays: days.length ? days : [a], time_local: ts[0], week_parity: "every" },
      ];
    }
  }
}

export function cycleIsExact(kind: CycleKind, meta: CycleMeta | undefined): boolean {
  if (kind !== "every_n_days") return true;
  return 14 % nValue(meta) === 0;
}

/** Whether ``d`` falls in a week matching the slot cycle (ISO calendar week). */
export function weekParityMatches(d: Date, parity: WeekParity): boolean {
  if (parity === "odd") return isoWeekNumber(d) % 2 === 1;
  if (parity === "even") return isoWeekNumber(d) % 2 === 0;
  return true;
}

export interface PreviewDay {
  date: Date;
  run: boolean;
  isToday: boolean;
}

/** 14-day strip starting at `start`; a day runs when any slot fires on it. */
export function previewStrip(
  slots: CycleSlotSpec[],
  start: Date,
  today: Date,
  days = 14
): PreviewDay[] {
  const out: PreviewDay[] = [];
  const tKey = today.toDateString();
  for (let i = 0; i < days; i++) {
    const d = new Date(start.getFullYear(), start.getMonth(), start.getDate() + i);
    const wd = mondayBasedWeekday(d);
    const run = slots.some((s) => s.weekdays.includes(wd) && weekParityMatches(d, s.week_parity));
    out.push({ date: d, run, isToday: d.toDateString() === tKey });
  }
  return out;
}

/** Day-gaps between consecutive fires over a 28-day window from `start`. */
export function previewGaps(slots: CycleSlotSpec[], start: Date, days = 28): number[] {
  const fires: Date[] = [];
  for (let i = 0; i < days; i++) {
    const d = new Date(start.getFullYear(), start.getMonth(), start.getDate() + i);
    const wd = mondayBasedWeekday(d);
    if (slots.some((s) => s.weekdays.includes(wd) && weekParityMatches(d, s.week_parity))) {
      fires.push(d);
    }
  }
  const gaps: number[] = [];
  for (let i = 0; i < fires.length - 1; i++) {
    gaps.push(Math.round((fires[i + 1].getTime() - fires[i].getTime()) / 86400000));
  }
  return gaps;
}

/** First date on/after `start` on which any slot fires (null if none within 28 days). */
export function firstRunDate(slots: CycleSlotSpec[], start: Date, days = 28): Date | null {
  for (let i = 0; i < days; i++) {
    const d = new Date(start.getFullYear(), start.getMonth(), start.getDate() + i);
    const wd = mondayBasedWeekday(d);
    if (slots.some((s) => s.weekdays.includes(wd) && weekParityMatches(d, s.week_parity))) {
      return d;
    }
  }
  return null;
}

File: custom_components/simple_irrigation/cycle.py
"""Cycle → schedule-slot generation (mirrored in frontend/src/cycle.ts).

A *cycle* is a user-facing watering cadence ("Every 2 days", "Twice daily", …).
It is pure presentation + generation metadata: the runtime only ever sees the
generated ``weekdays`` / ``time_local`` / ``week_parity`` on each member slot.

Cycle length is 7 or 14 days because ``week_parity`` can only express a 2-week
cycle. All rules here MUST match ``cycle.ts`` exactly so the wizard preview and
the created slots agree 100 %.
"""

from __future__ import annotations

import math
from datetime import date, timedelta
from typing import Any

from .const import WEEK_PARITY_EVEN, WEEK_PARITY_EVERY, WEEK_PARITY_ODD

CYCLE_KINDS = (
    "daily",
    "twice_daily",
    "every_n_days",
    "n_per_week",
    "weekly",
    "biweekly",
    "custom",
)


def round_half_up(x: float) -> int:
    """Round half away from zero-ish (half up), matching JS ``Math.round``.

    Python's built-in ``round`` uses banker's rounding, which diverges from the
    frontend for values like 10.5; use this everywhere offsets are computed.
    """
    return math.floor(x + 0.5)


def opposite_parity(parity: str) -> str:
    """Flip odd<->even; ``every`` stays ``every``."""
    if parity == WEEK_PARITY_ODD:
        return WEEK_PARITY_EVEN
    if parity == WEEK_PARITY_EVEN:
        return WEEK_PARITY_ODD
    return WEEK_PARITY_EVERY


def anchor_week_parity(anchor_weekday: int, today: date) -> str:
    """Parity (odd/even) of the ISO week containing the next occurrence of the anchor."""
    a = max(0, min(6, int(anchor_weekday)))
    for i in range(8):
        d = today + timedelta(days=i)
        if d.weekday() == a:
            return WEEK_PARITY_ODD if d.isocalendar()[1] % 2 == 1 else WEEK_PARITY_EVEN
    return WEEK_PARITY_EVERY


def _times(meta: dict[str, Any] | None) -> list[str]:
    raw = (meta or {}).get("times")
    out: list[str] = []
    if isinstance(raw, (list, tuple)):
        for x in raw:
            s = str(x).strip()
            if s:
                out.append(s)
    if not out:
        out = ["06:00"]
    return out


def _anchor(meta: dict[str, Any] | None) -> int:
    try:
        return max(0, min(6, int((meta or {}).get("anchor_weekday", 0))))
    except (TypeError, ValueError):
        return 0


def _n(meta: dict[str, Any] | None, default: int = 2) -> int:
    try:
        return max(1, int((meta or {}).get("n", default)))
    except (TypeError, ValueError):
        return default


def _week_days(meta: dict[str, Any] | None) -> list[int]:
    raw = (meta or {}).get("week_days")
    out: list[int] = []
    seen: set[int] = set()
    if isinstance(raw, (list, tuple, set)):
        for x in raw:
            try:
                v = int(x)
            except (TypeError, ValueError):
                continue
            if 0 <= v <= 6 and v not in seen:
                seen.add(v)
                out.append(v)
    out.sort()
    return out


def _every_n_days_slots(n: int, anchor: int, time_local: str, p0: str) -> list[dict[str, Any]]:
    """Split an every-N-days cadence over a 7/14-day cycle into slot specs."""
    runs = max(1, round_half_up(14 / n))
    offsets = [round_half_up(k * 14 / runs) for k in range(runs)]

    week_a: list[int] = []
    week_b: list[int] = []
    seen_a: set[int] = set()
    seen_b: set[int] = set()
    for off in offsets:
        wd = (anchor + off) % 7
        if off < 7:
            if wd not in seen_a:
                seen_a.add(wd)
                week_a.append(wd)
        else:
            if wd not in seen_b:
                seen_b.add(wd)
                week_b.append(wd)
    week_a.sort()
    week_b.sort()

    if week_a and week_b and seen_a == seen_b:
        return [{"weekdays": week_a, "time_local": time_local, "week_parity": WEEK_PARITY_EVERY}]

    slots: list[dict[str, Any]] = []
    if week_a:
        slots.append({"weekdays": week_a, "time_local": time_local, "week_parity": p0})
    if week_b:
        slots.append(
            {
                "weekdays": week_b,
                "time_local": time_local,
                "week_parity": opposite_parity(p0),
            }
        )
    return slots


def generate_cycle_slots(
    kind: str,
    meta: dict[str, Any] | None,
    *,
    anchor_parity: str = WEEK_PARITY_ODD,
) -> list[dict[str, Any]]:
    """Return slot specs ``[{weekdays, time_local, week_parity}, …]`` for a cycle.

    ``anchor_parity`` (``P0``) is the parity of the ISO week containing the next
    occurrence of the anchor weekday; only ``biweekly`` and the two-slot
    ``every_n_days`` case use it.
    """
    times = _times(meta)
    anchor = _anchor(meta)
    all_days = [0, 1, 2, 3, 4, 5, 6]

    if kind == "daily":
        return [{"weekdays": all_days, "time_local": times[0], "week_parity": WEEK_PARITY_EVERY}]

    if kind == "twice_daily":
        t2 = times[1] if len(times) > 1 else times[0]
        return [
            {"weekdays": all_days, "time_local": times[0], "week_parity": WEEK_PARITY_EVERY},
            {"weekdays": all_days, "time_local": t2, "week_parity": WEEK_PARITY_EVERY},
        ]

    if kind == "weekly":
        return [{"weekdays": [anchor], "time_local": times[0], "week_parity": WEEK_PARITY_EVERY}]

    if kind == "biweekly":
        return [{"weekdays": [anchor], "time_local": times[0], "week_parity": anchor_parity}]

    if kind == "n_per_week":
        days = _week_days(meta) or [anchor]
        return [{"weekdays": days, "time_local": times[0], "week_parity": WEEK_PARITY_EVERY}]

    if kind == "every_n_days":
        return _every_n_days_slots(_n(meta), anchor, times[0], anchor_parity)

    # custom (or unknown): a single slot from the chosen weekdays.
    days = _week_days(meta) or [anchor]
    return [{"weekdays": days, "time_local": times[0], "week_parity": WEEK_PARITY_EVERY}]


def cycle_is_exact(kind: str, meta: dict[str, Any] | None) -> bool:
    """Whether the cadence divides the 14-day cycle evenly (no ``approx`` badge)."""
    if kind != "every_n_days":
        return True
    return 14 % _n(meta) == 0


def simulate_fire_days(
    slots: list[dict[str, Any]], start: date, days: int
) -> list[date]:
    """Dates on which any slot fires in ``[start, start+days)`` (runtime semantics)."""
    from .time_util import week_parity_matches

    fires: list[date] = []
    for i in range(days):
        d = start + timedelta(days=i)
        for s in slots:
            if d.weekday() not in s["weekdays"]:
                continue
            if not week_parity_matches(d, s["week_parity"]):
                continue
            fires.append(d)
            break
    return fires


def cycle_gaps(
    kind: str, meta: dict[str, Any] | None, start: date, anchor_parity: str = WEEK_PARITY_ODD
) -> list[int]:
    """Real day-gap sequence between consecutive fires over a two-cycle window."""
    slots = generate_cycle_slots(kind, meta, anchor_parity=anchor_parity)
    fires = simulate_fire_days(slots, start, 28)
    if len(fires) < 2:
        return []
    return [(fires[i + 1] - fires[i]).days for i in range(len(fires) - 1)]
